invesalius/data/log.py

```diff
@@ -0,0 +1,295 @@
+#    Este programa e software livre; voce pode redistribui-lo e/ou
+#    modifica-lo sob os termos da Licenca Publica Geral GNU, conforme
+#    publicada pela Free Software Foundation; de acordo com a versao 2
+#    da Licenca.
+#
+#    Este programa eh distribuido na expectativa de ser util, mas SEM
+#    QUALQUER GARANTIA; sem mesmo a garantia implicita de
+#    COMERCIALIZACAO ou de ADEQUACAO A QUALQUER PROPOSITO EM
+#    PARTICULAR. Consulte a Licenca Publica Geral GNU para obter mais
+#    detalhes.
+#-------------------------------------------------------------------------
+
+import logging 
+import logging.config 
+from typing import Callable
+import sys, os
+import wx
+import json
+
+from datetime import datetime
+
+from invesalius import inv_paths
+import invesalius.constants as const
+from invesalius.utils import Singleton, deep_merge_dict
+
+LOG_CONFIG_PATH = os.path.join(inv_paths.USER_INV_DIR, 'log_config.json')
+DEFAULT_LOGFILE = os.path.join(inv_paths.USER_LOG_DIR,datetime.now().strftime("invlog-%Y_%m_%d-%I_%M_%S_%p.log"))
+
+class MyConsoleHandler(logging.StreamHandler):
+    def __init__(self, textctrl):
+        logging.StreamHandler.__init__(self)
+        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        self.setFormatter(formatter)
+        self.textctrl = textctrl
+
+    def emit(self, record):
+        msg = self.format(record)
+        stream = self.stream
+        self.textctrl.WriteText(msg + "\n")
+        self.flush()
+
+class MyRedirectText(object):
+    def __init__(self, textctrl):
+        self.out = textctrl
+        
+    def write(self, string):
+        self.out.WriteText(string)
+
+class MyPanel(wx.Panel):
+    def __init__(self, parent, logger):
+        wx.Panel.__init__(self, parent)
+        
+        logText = wx.TextCtrl(self, style = wx.TE_MULTILINE|wx.TE_READONLY|wx.HSCROLL)
+        
+        sizer = wx.BoxSizer(wx.VERTICAL)
+        sizer.Add(logText, 1, wx.EXPAND|wx.ALL, 5)
+        #sizer.Add(btn, 0, wx.ALL, 5)
+        self.SetSizer(sizer)
+        self._parent = parent
+     
+        redir = MyRedirectText(logText)
+        sys.stdout = redir
+        #sys.stderr = redir
+
+        self._logger = logger
+        txtHandler = MyConsoleHandler(logText)
+        self._logger.addHandler(txtHandler)
+
+    def onClose(self, event):
+        self._logger.info("MyPanel window close selected.")
+        self._parent.Iconify()
+        
+     
+class MyFrame(wx.Frame):
+    def __init__(self, logger):
+        wx.Frame.__init__(self, None, title="Log Console", 
+                    style=wx.DEFAULT_FRAME_STYLE & (~wx.CLOSE_BOX) & (~wx.MAXIMIZE_BOX))
+        self._panel = MyPanel(self, logger)
+        self.Bind(wx.EVT_CLOSE, self.onClose)
+        self.Show()
+        
+    #To be tested
+    def onClose(self, event):
+        self._logger.info("MyFrame window close selected.")
+        self.Hide()
+
+class MyLogger(metaclass=Singleton):
```

Comment:
Does it need to be a singleton?

Reply:
Not so. Python "Logging" is itself a Singleton. I will test it and make changes.

Reply:
Done

Reply:
I have a question though. Right now, my logger class is not a singleton, and I have an instance of the class placed in log.py which I plan to access from everywhere. I am not sure if the log.py file is imported from two locations. In that case, I may have to revert back to a singleton. Please comment.